python/cucim/src/cucim/skimage/filters/_median_hist.py

```diff
@@ -0,0 +1,733 @@
+import math
+from collections import namedtuple
+from textwrap import dedent
+
+import cupy as cp
+import numpy as np
+
+from .._shared.utils import _to_np_mode
+
+if hasattr(math, 'prod'):
+    prod = math.prod
+else:
+    prod = np.prod
+
+
+def _dtype_to_CUDA_int_type(dtype):
+    cpp_int_types = {
+        cp.uint8: 'unsigned char',
+        cp.uint16: 'unsigned short',
+        cp.uint32: 'unsigned int',
+        cp.uint64: 'unsigned long long',
+        cp.int8: 'signed char',
+        cp.int16: 'short',
+        cp.int32: 'int',
+        cp.int64: 'long long',
+    }
+    dtype = cp.dtype(dtype)
+    if dtype.type not in cpp_int_types:
+        raise ValueError(f"unrecognized dtype: {dtype.type}")
+    return cpp_int_types[dtype.type]
+
+
+def _get_hist_dtype(footprint_shape):
+    """Determine C++ type and cupy.dtype to use for the histogram."""
+    max_possible_count = prod(footprint_shape)
+
+    if max_possible_count < 128:
+        dtype = cp.int8
+    elif max_possible_count < 32768:
+        dtype = cp.int16
+    else:
+        dtype = cp.int32
+    return _dtype_to_CUDA_int_type(dtype), dtype
+
+
+def _gen_global_definitions(
+    image_t='unsigned char',
+    hist_offset=0,
+    hist_int_t='int',
+    hist_size=256,
+    hist_size_coarse=8
+):
+    """Generate C++ #define statements needed for the CUDA kernels.
+
+    The definitions used depend on the number of histogram bins and the
+    histogram data type.
+    """
+
+    if hist_size % hist_size_coarse != 0:
+        raise ValueError(
+            "`hist_size` must be a multiple of `hist_size_coarse`"
+        )
+    hist_size_fine = hist_size // hist_size_coarse
+    log2_coarse = math.log2(hist_size_coarse)
+    log2_fine = math.log2(hist_size_fine)
+    if abs(math.remainder(log2_coarse, 1)) > 1e-7:
+        raise ValueError("log2_coarse must be a power of two")
+    elif abs(math.remainder(log2_fine, 1)) > 1e-7:
+        raise ValueError("log2_fine must be a power of two")
+    else:
+        log2_coarse = round(log2_coarse)
+        log2_fine = round(log2_fine)
+
+    global_defs = f"""
+#define HIST_SIZE {hist_size}
+#define HIST_SIZE_COARSE {hist_size_coarse}
+#define HIST_SIZE_FINE {hist_size_fine}
+#define HIST_INT_T {hist_int_t}
+#define HIST_OFFSET {hist_offset}
+#define IMAGE_T {image_t}
+#define LOG2_COARSE {log2_coarse}
+#define LOG2_FINE {log2_fine}
+    """
+    return global_defs
+
+
+# This preamble code is common across all generated kernels
+preamble_common = """
+
+__device__ void histogramAddAndSubCoarse(HIST_INT_T* H, const HIST_INT_T* hist_colAdd, const HIST_INT_T* hist_colSub){
+    int tx = threadIdx.x;
+    if (tx<HIST_SIZE_COARSE){
+        H[tx]+=hist_colAdd[tx]-hist_colSub[tx];
+    }
+}
```

Comment:
Maybe it is worth pulling this code out into C files that we load and then compile?

Reply:
That was the route I ended up going with the distance transform PR. We could do something similar here.

Reply:
Please see the change in [2608b2e](https://github.com/rapidsai/cucim/pull/317/commits/2608b2e9771ea4143f10ca8c29660cc0a19150ca) and [a364ee4](https://github.com/rapidsai/cucim/pull/317/commits/a364ee46e90bb16531fc3b0b1f21f4948413bb6c) where I moved all of the non-dynamically generated code into a separate .cu file.

Reply:
Thanks Greg! 🙏

Think we may want to start doing this in more places going forward. Filed issue ( https://github.com/rapidsai/cucim/issues/375 ) about it